### PR TITLE
Remove DiagnosticBuilder.quiet

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @eddyb 

cc https://github.com/rust-lang/rust/issues/69426 @GuillaumeGomez @Manishearth

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Like I said, please provide an alternative before removing this.

**Comment by GuillaumeGomez** [comment]
Talked with @oli-obk and the simplest solution is to just fix errors reported by the compiler (some missing `cfg` in the `std` modules). All good for me then!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by eddyb** [comment]
If it's an acceptable tradeoff, I would use the typical pattern of requesting `TypeckResults` when visiting each HIR body, effectively switching the `rustdoc` default of omitting `typeck` on anything that isn't requested by something else (e.g. a `const fn` being called from an array length).

It might even be easier to not engage in those name resolution / `typeck` omissions in the first place, if it produces better errors overall.
So basically when this annotating feature is being used, `rustdoc` would stop being "lenient" with regards to `cfg`'d out code, and start being "strict", in order to have as much information as possible.

@bors r+

**Comment by Manishearth** [comment]
> Like I said, please provide an alternative before removing this.

@GuillaumeGomez my understanding is that this feature is causing soundness issues (because it turns errors _off_ completely). From @eddyb's reaction when talking to me about it it seems like this very much should never have happened.

As such I see zero reason why an experimental rustdoc feature should delay the fixing of such a bug.

**Comment by mark-i-m** [comment]
@Manishearth Perhaps you (like me) missed their most recent message?

**Comment by Manishearth** [comment]
@mark-i-m No, I saw the updates (and there's a line in my comment acknowledging the issue got resolved already), from the rustdoc side I'm very strongly against bugs in this feature blocking soundness fixes (or really, any compiler work) and want to push back against our team adopting such an attitude.

**Comment by GuillaumeGomez** [comment]
I don't think what I asked was unreasonable. This has been around for months and it seems normal for me to worry about how to go around this limitation.

> No, I saw the updates (and there's a line in my comment acknowledging the issue got resolved already), from the rustdoc side I'm very strongly against bugs in this experimental feature blocking soundness fixes (or really, any compiler work) and want to push back against our team adopting such an attitude.

Humans make mistakes. Please scold me in private next time. There is nothing about "such an attitude", just a legitimate concern.

**Comment by eddyb** [comment]
> This has been around for months 

To be fair, it's been around for months with `rustc` changes unreviewed by @rust-lang/compiler - I haven't been that active and only noticed because @mark-i-m pinged me on a PR that seems unsound (or rather, reveals pre-existing risks) in the presence of such a mechanism to silence even the compiler-observable effects of errors (e.g. `ty::Error`). It could've been an year later, if nobody noticed.
The situation is similar to #59789, and there might be others. I'm not sure what to do, other than impose stricter review requirements for changes in the `compiler` directory but... that's not also foolproof.

**Comment by GuillaumeGomez** [comment]
As far as I know it's only used in rustdoc but I understand your problem with it. As for this, since it was in the error system, I thought that @estebank was the one responsible. Sorry for the noise.
